File: app/bot/api_client.py
"""
Zentra API Client — bot uchun backend FastAPI bilan ishlaydigan HTTP klient.

Dizayn qoidasi: bot ichida hech qanday biznes-mantiq (soliq, hisob-kitob,
ruxsatlar) YO'Q — bu klient faqat HTTP so'rov yuboradi va javobni qaytaradi.
Access token muddati o'tgan bo'lsa, avtomatik ravishda refresh qilib,
so'rovni bir marta qayta uradi.
"""

from __future__ import annotations

from typing import Any

import httpx
from loguru import logger

from app.bot.session_store import BotSession, session_store
from app.core.config import settings


class APIError(Exception):
    """Backend 2xx/3xx dan boshqa status qaytarganda ko'tariladi."""

    def __init__(self, status_code: int, detail: str) -> None:
        self.status_code = status_code
        self.detail = detail
        super().__init__(f"[{status_code}] {detail}")


class AuthRequiredError(Exception):
    """Sessiya yo'q yoki refresh ham muvaffaqiyatsiz bo'lganda — foydalanuvchi qayta /start bosishi kerak."""


class ZentraAPIClient:
    def __init__(self, base_url: str | None = None) -> None:
        self._base_url = (base_url or settings.API_BASE_URL).rstrip("/")
        self._prefix = settings.API_V1_PREFIX

    def _url(self, path: str) -> str:
        return f"{self._base_url}{self._prefix}{path}"

    # ─── Autentifikatsiya (token talab qilmaydi) ─────────────────────────────
    async def telegram_login(self, telegram_id: int, full_name: str, language: str = "uz") -> dict:
        async with httpx.AsyncClient(timeout=15.0) as client:
            resp = await client.post(
                self._url("/auth/telegram"),
                json={"telegram_id": telegram_id, "full_name": full_name, "language": language},
            )
            self._raise_for_status(resp)
            return resp.json()

    async def _refresh(self, refresh_token: str) -> dict:
        async with httpx.AsyncClient(timeout=15.0) as client:
            resp = await client.post(
                self._url("/auth/refresh"),
                json={"refresh_token": refresh_token},
            )
            self._raise_for_status(resp)
            return resp.json()

    # ─── Umumiy so'rov (token bilan, kerak bo'lsa avtomatik refresh) ────────
    async def request(
        self,
        telegram_id: int,
        method: str,
        path: str,
        *,
        json_body: dict | None = None,
        params: dict | None = None,
    ) -> Any:
        session = await session_store.get(telegram_id)
        if session is None:
            raise AuthRequiredError("Sessiya topilmadi — /start bosing")

        response = await self._do_request(session.access_token, method, path, json_body, params)

        if response.status_code == 401:
            # Access token eskirgan bo'lishi mumkin — refresh qilib qayta urinamiz
            try:
                tokens = await self._refresh(session.refresh_token)
            except APIError:
                await session_store.delete(telegram_id)
                raise AuthRequiredError("Sessiya muddati tugagan — /start bosing") from None

            session.access_token = tokens["access_token"]
            session.refresh_token = tokens["refresh_token"]
            await session_store.set(session)

            response = await self._do_request(session.access_token, method, path, json_body, params)

        self._raise_for_status(response)
        if response.status_code == 204 or not response.content:
            return None

        content_type = response.headers.get("content-type", "")
        if "application/json" in content_type:
            return response.json()
        return response.content  # PDF/Excel bayt oqimi

    async def _do_request(
        self,
        access_token: str,
        method: str,
        path: str,
        json_body: dict | None,
        params: dict | None,
    ) -> httpx.Response:
        async with httpx.AsyncClient(timeout=30.0) as client:
            return await client.request(
                method,
                self._url(path),
                json=json_body,
                params=params,
                headers={"Authorization": f"Bearer {access_token}"},
            )

    @staticmethod
    def _raise_for_status(response: httpx.Response) -> None:
        if response.status_code >= 400:
            try:
                detail = response.json().get("detail", response.text)
            except ValueError:
                detail = response.text
            logger.warning(f"API xato: {response.status_code} {response.request.url} | {detail}")
            raise APIError(response.status_code, str(detail))

    # ─── Qulaylik metodlari (ko'p ishlatiladigan endpointlar) ───────────────
    async def get_me(self, telegram_id: int) -> dict:
        return await self.request(telegram_id, "GET", "/auth/me")

    async def list_companies(self, telegram_id: int) -> list[dict]:
        return await self.request(telegram_id, "GET", "/companies")

    async def create_company(self, telegram_id: int, name: str, **extra: Any) -> dict:
        return await self.request(telegram_id, "POST", "/companies", json_body={"name": name, **extra})

    async def ai_chat(
        self, telegram_id: int, company_id: str, message: str, session_id: str | None = None
    ) -> dict:
        return await self.request(
            telegram_id,
            "POST",
            f"/ai/{company_id}/chat",
            json_body={"message": message, "session_id": session_id},
        )

    async def ai_ocr(self, telegram_id: int, company_id: str, image_base64: str, media_type: str) -> dict:
        return await self.request(
            telegram_id,
            "POST",
            f"/ai/{company_id}/ocr",
            json_body={"image_base64": image_base64, "media_type": media_type},
        )

    async def list_transactions(self, telegram_id: int, company_id: str, limit: int = 10) -> list[dict]:
        return await self.request(
            telegram_id, "GET", f"/transactions/{company_id}", params={"limit": limit}
        )

    async def create_transaction(self, telegram_id: int, company_id: str, **body: Any) -> dict:
        return await self.request(telegram_id, "POST", f"/transactions/{company_id}", json_body=body)

    async def confirm_transaction(self, telegram_id: int, company_id: str, transaction_id: str) -> dict:
        return await self.request(
            telegram_id, "PATCH", f"/transactions/{company_id}/{transaction_id}/confirm"
        )

    async def dashboard(self, telegram_id: int, company_id: str, period_start: str, period_end: str) -> dict:
        return await self.request(
            telegram_id,
            "GET",
            f"/reports/{company_id}/dashboard",
            params={"period_start": period_start, "period_end": period_end},
        )

    async def list_debts(self, telegram_id: int, company_id: str, overdue_only: bool = False) -> list[dict]:
        path = f"/debts/{company_id}/overdue" if overdue_only else f"/debts/{company_id}"
        return await self.request(telegram_id, "GET", path)

    async def create_debt(self, telegram_id: int, company_id: str, **body: Any) -> dict:
        return await self.request(telegram_id, "POST", f"/debts/{company_id}", json_body=body)

    async def list_customers(self, telegram_id: int, company_id: str, search: str | None = None) -> list[dict]:
        params = {"search": search} if search else None
        return await self.request(telegram_id, "GET", f"/customers/{company_id}", params=params)

    async def create_customer(self, telegram_id: int, company_id: str, name: str, **extra: Any) -> dict:
        return await self.request(
            telegram_id, "POST", f"/customers/{company_id}", json_body={"name": name, **extra}
        )

    async def list_suppliers(self, telegram_id: int, company_id: str, search: str | None = None) -> list[dict]:
        params = {"search": search} if search else None
        return await self.request(telegram_id, "GET", f"/suppliers/{company_id}", params=params)

    async def create_supplier(self, telegram_id: int, company_id: str, name: str, **extra: Any) -> dict:
        return await self.request(
            telegram_id, "POST", f"/suppliers/{company_id}", json_body={"name": name, **extra}
        )

    async def debts_aging(self, telegram_id: int, company_id: str) -> dict:
        return await self.request(telegram_id, "GET", f"/debts/{company_id}/aging")

    async def upcoming_tax_deadlines(self, telegram_id: int, days_ahead: int = 30) -> dict:
        return await self.request(
            telegram_id, "GET", "/tax/calendar/upcoming", params={"days_ahead": days_ahead}
        )

    async def calculate_vat(self, telegram_id: int, company_id: str, amount: float) -> dict:
        return await self.request(
            telegram_id, "POST", f"/tax/{company_id}/calculate-vat", json_body={"amount": amount}
        )

    async def generate_invoice(self, telegram_id: int, company_id: str, transaction_id: str) -> bytes:
        return await self.request(telegram_id, "GET", f"/documents/{company_id}/invoice/{transaction_id}")

    async def generate_act(self, telegram_id: int, company_id: str, transaction_id: str) -> bytes:
        return await self.request(telegram_id, "GET", f"/documents/{company_id}/act/{transaction_id}")

    async def export_transactions_excel(self, telegram_id: int, company_id: str) -> bytes:
        return await self.request(telegram_id, "GET", f"/documents/{company_id}/export/transactions")


api_client = ZentraAPIClient()

File: frontend/src/lib/apiClient.js
/**
 * Zentra API klienti.
 *
 * Bot'dagi app/bot/api_client.py bilan bir xil g'oya: access token muddati
 * o'tsa (401), refresh token bilan avtomatik yangilaymiz va so'rovni bir
 * marta qayta uramiz. Tokenlar localStorage'da saqlanadi.
 */

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || "http://localhost:8000";
const API_PREFIX = "/api/v1";

const TOKEN_KEY = "zentra.tokens";

export class APIError extends Error {
  constructor(status, detail) {
    super(typeof detail === "string" ? detail : JSON.stringify(detail));
    this.status = status;
    this.detail = detail;
  }
}

export function getStoredTokens() {
  const raw = localStorage.getItem(TOKEN_KEY);
  if (!raw) return null;
  try {
    return JSON.parse(raw);
  } catch {
    localStorage.removeItem(TOKEN_KEY);
    return null;
  }
}

export function storeTokens(tokens) {
  localStorage.setItem(TOKEN_KEY, JSON.stringify(tokens));
}

export function clearTokens() {
  localStorage.removeItem(TOKEN_KEY);
}

async function rawRequest(path, { method = "GET", body, token, params } = {}) {
  const url = new URL(`${API_BASE_URL}${API_PREFIX}${path}`);
  if (params) {
    Object.entries(params).forEach(([k, v]) => {
      if (v !== undefined && v !== null) url.searchParams.set(k, v);
    });
  }

  const headers = { "Content-Type": "application/json" };
  if (token) headers.Authorization = `Bearer ${token}`;

  const response = await fetch(url, {
    method,
    headers,
    body: body !== undefined ? JSON.stringify(body) : undefined,
  });

  return response;
}

async function refreshTokens() {
  const tokens = getStoredTokens();
  if (!tokens?.refresh_token) throw new APIError(401, "Sessiya yo'q");

  const response = await rawRequest("/auth/refresh", {
    method: "POST",
    body: { refresh_token: tokens.refresh_token },
  });

  if (!response.ok) {
    clearTokens();
    throw new APIError(response.status, "Sessiya muddati tugagan");
  }

  const newTokens = await response.json();
  storeTokens(newTokens);
  return newTokens;
}

/**
 * Asosiy so'rov funksiyasi — token talab qiladigan barcha endpointlar shu orqali chaqiriladi.
 * 401 kelsa: bitta marta refresh qilib qayta uradi.
 */
export async function request(path, options = {}) {
  let tokens = getStoredTokens();
  if (!tokens?.access_token) {
    throw new APIError(401, "Kirilmagan — /login sahifasiga o'ting");
  }

  let response = await rawRequest(path, { ...options, token: tokens.access_token });

  if (response.status === 401) {
    tokens = await refreshTokens();
    response = await rawRequest(path, { ...options, token: tokens.access_token });
  }

  if (!response.ok) {
    let detail = response.statusText;
    try {
      const errBody = await response.json();
      detail = errBody.detail || detail;
    } catch {
      /* javob JSON emas — statusText bilan qolamiz */
    }
    throw new APIError(response.status, detail);
  }

  if (response.status === 204) return null;

  const contentType = response.headers.get("content-type") || "";
  if (contentType.includes("application/json")) return response.json();
  return response.blob(); // PDF/Excel
}

/** Autentifikatsiyasiz so'rov (login, register, widget auth uchun). */
export async function publicRequest(path, options = {}) {
  const response = await rawRequest(path, options);
  if (!response.ok) {
    let detail = response.statusText;
    try {
      const errBody = await response.json();
      detail = errBody.detail || detail;
    } catch {
      /* noop */
    }
    throw new APIError(response.status, detail);
  }
  return response.json();
}

// ─── Qulaylik metodlari ──────────────────────────────────────────────────
export const api = {
  // Auth
  loginWithTelegramWidget: (widgetData) =>
    publicRequest("/auth/telegram-widget", { method: "POST", body: widgetData }),
  me: () => request("/auth/me"),

  // Companies
  listCompanies: () => request("/companies"),
  createCompany: (body) => request("/companies", { method: "POST", body }),

  // Dashboard / Reports
  dashboard: (companyId, periodStart, periodEnd) =>
    request(`/reports/${companyId}/dashboard`, {
      params: { period_start: periodStart, period_end: periodEnd },
    }),

  // Transactions
  listTransactions: (companyId, limit = 20) =>
    request(`/transactions/${companyId}`, { params: { limit } }),
  confirmTransaction: (companyId, txId) =>
    request(`/transactions/${companyId}/${txId}/confirm`, { method: "PATCH" }),

  // Debts
  listDebts: (companyId, overdueOnly = false) =>
    request(overdueOnly ? `/debts/${companyId}/overdue` : `/debts/${companyId}`),
  debtsAging: (companyId) => request(`/debts/${companyId}/aging`),
  createDebt: (companyId, body) => request(`/debts/${companyId}`, { method: "POST", body }),

  // Customers / Suppliers
  listCustomers: (companyId, search) =>
    request(`/customers/${companyId}`, { params: { search } }),
  createCustomer: (companyId, body) => request(`/customers/${companyId}`, { method: "POST", body }),
  listSuppliers: (companyId, search) =>
    request(`/suppliers/${companyId}`, { params: { search } }),
  createSupplier: (companyId, body) => request(`/suppliers/${companyId}`, { method: "POST", body }),

  // Tax
  upcomingTaxDeadlines: (daysAhead = 30) =>
    request("/tax/calendar/upcoming", { params: { days_ahead: daysAhead } }),
};
